Bluetooth: host: direction: Add public API to set CTE TX enable for adv.

ppryga-nordic opened the pull request.

ppryga-nordic committed:
Bluetooth: controller: df: Config help corrction correction

Correct names of kconfig entries to mach to style of other entries.
Style correction in hci.c. Use the same approach of checking if
Kconfig entry is enabled.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: direction: Fix compilation error

Fix compilation error caused by missing header files.
Reorganized order of header files include.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/direction.c:
```diff
@@ -155,6 +156,47 @@ static int hci_df_read_ant_info(uint8_t *switch_sample_rates,
+			   enable);
+	cmd_data_state_set(buf, &state);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_CL_CTE_TX_ENABLE,
+				   buf, NULL);
+
+	if (err) {
+		return err;
+	}
+
+	return 0;
```
I was first going to point out to remove the empty line after assignment to `err` and checking its value, but if you're not going to extend this in the future you could just eliminate `err` completely and do `return bt_hci_cmd_send_sync(...);`

ppryga replied:
Corrected according to suggestion

jhedberg commented on subsys/bluetooth/host/direction.c:
```diff
@@ -270,10 +312,53 @@ int bt_df_set_adv_cte_tx_param(struct bt_le_ext_adv *adv,
+		return -EALREADY;
+	}
+
+	err = hci_df_set_adv_cte_tx_enable(adv, enable);
+	if (err) {
+		return err;
+	}
+
 	return 0;
```
Same here, since you're not really doing anything with `err`

ppryga replied:
Corrected according to suggestion

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -305,6 +313,29 @@ extern const struct bt_conn_auth_cb *bt_auth;
+	state->target = target;
+	state->bit = bit;
+	state->val = val;
+}
+/* Set commnad state for given network buffer */
```
Missing empty line

ppryga replied:
Added empty line before comment line.

jhedberg reviewed (changes requested):
Looks generally ok, just a couple of style comments

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -305,6 +313,30 @@ extern const struct bt_conn_auth_cb *bt_auth;
+}
+
+/* Set commnad state for given network buffer */
+void cmd_data_state_set(struct net_buf *buf, struct cmd_state_set *state);
```
Sorry, I should have commented on this as well. Even though it's an internal header file, `hci_core.h` has had a strict `bt_*` namespace so far, and it seems you're the first one to try to introduce content to it that violates that. It's not really a big deal since these are internal files, but it will cause some confusion when starting to look for the definitions, since when I don't see a `bt_*` prefix I'd assume that they are located in the same c-file.

ppryga replied:
Changed as suggested. I thought that only public API has bt_ prefix. What about hci_ (e.g. hci_evt_pin_code_req)? Is it an exception for HCI event handlers?

jhedberg replied:
If you look at the header files under `subsys/bluetooth/host/*.h` (which are all internal, public files are found under `include/bluetooth`) you'll see that pretty much everything follows the `bt_*` namespace

jhedberg replied:
We can argue about whether it makes sense, but the higher priority IMO is to keep things consistent

ppryga replied:
I was not going to argue ;-) I was just lecturing about conventions in Host implementation.

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -305,6 +313,30 @@ extern const struct bt_conn_auth_cb *bt_auth;
+/* Data type to store state related with commnad to be updated
+ * when command completes successfully.
+ */
+struct cmd_state_set {
```
And yes, I'm not really happy about long symbol names, but consistency trumps that

jhedberg replied:
Yes, use the same prefix for both

ppryga replied:
It is ready

cvinayak reviewed (approved):
Reviewed controller changes, LGTM.

jhedberg commented:
@ppryga a couple of nitpicks:

 - Remove `.` from the end of commit message summaries. Now two have it and two don't.
 - Split "Bluetooth: host: direction: Add public API to set CTE TX enable for adv" into two commits: one which refactors and renames the state stuff, and another for the new additions. @joerchan btw it'd be good to get a review from you since IIRC you implemented the state stuff.

ppryga-nordic committed:
Bluetooth: host: Rename and make non-static HCI cmd state update

Enable CTE HCI command requires update of a state of its
execution after receive command completion response from
controller. Host has already implemented such functionality
but it was used only internally by hci_core.

This commit changes it to be accessible from other sourece files
in host.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: hci: Add HCI_Set_Connectionless_CTE_TX_Enable cmd

Add implementation of HCI_Set_Connecitonless_CTE_TX_Enable
command to HCI. Add scratch implementation of command handling
functions to controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: direction: Add public API to set CTE TX enable for adv

Add public function to set Constant Tone Extension transmission
enabled or disabled for periodic advertising.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
@jhedberg requested updates are done.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -37,4 +37,28 @@ struct bt_df_adv_cte_tx_param {
 int bt_df_set_adv_cte_tx_param(struct bt_le_ext_adv *adv,
 			       const struct bt_df_adv_cte_tx_param *params);
 
+/**
```
Please start with brief on first line `/** @breif ...` for consistency.

ppryga replied:
Corrected as requested

joerchan commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -303,6 +303,9 @@ uint8_t ll_df_set_cl_cte_tx_params(uint8_t adv_handle, uint8_t cte_len,
 				   uint8_t cte_type, uint8_t cte_count,
 				   uint8_t num_ant_ids, uint8_t *ant_ids);
 
+/* Enables of disables CTE TX for periodic advertising */
```
of -> or

ppryga replied:
Corrected

joerchan commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -193,6 +193,18 @@ uint8_t ll_df_set_cl_cte_tx_params(uint8_t adv_handle, uint8_t cte_len,
 	return BT_HCI_ERR_SUCCESS;
 }
 
+/* @brief Function enables of disables CTE TX for periodic advertising.
```
Enables or disables.

ppryga replied:
I'll leave it as is. Every other @brief in this file start with "Function", so I want to be consistent.

joerchan replied:
I was correcting the spelling: of -> or.

ppryga replied:
Sorry for misunderstanding, I've corrected it.

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -155,6 +156,40 @@ static int hci_df_read_ant_info(uint8_t *switch_sample_rates,
+static int hci_df_set_adv_cte_tx_enable(struct bt_le_ext_adv *adv,
+					bool enable)
+{
+	struct bt_hci_cp_le_set_per_adv_recv_enable *cp;
```
bt_hci_cp_le_set_cl_cte_tx_enable

ppryga replied:
Oh... sorry, how did it happen... Corrected to `bt_hci_cp_le_set_cl_cte_tx_enable`.

joerchan commented on subsys/bluetooth/host/direction.c:
```diff
@@ -155,6 +156,40 @@ static int hci_df_read_ant_info(uint8_t *switch_sample_rates,
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = adv->handle;
+	cp->enable = enable ? 1 : 0;
```
cte_enable

ppryga replied:
Corrected to `cte_enable`

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -157,6 +143,12 @@ static struct cmd_data cmd_data[CONFIG_BT_HCI_CMD_COUNT];
 #define cmd(buf) (&cmd_data[net_buf_id(buf)])
 #define acl(buf) ((struct acl_data *)net_buf_user_data(buf))
 
+void bt_hci_cmd_data_state_set(struct net_buf *buf,
```
maybe static inline this one in the header as well?

ppryga replied:
But then I also have to move lines 127-143 including definition of a static array cmd_data[]. That would not work because of multiple instances of the array variable in different translation units. Function goal is to give access to the array from other source files than hci_core.c.

joerchan replied:
Fair point :)

joerchan commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -305,6 +313,32 @@ extern const struct bt_conn_auth_cb *bt_auth;
+	bool val;
+};
+
+/* Initialize commnad state instance */
```
command

ppryga replied:
Corrected

joerchan commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -305,6 +313,32 @@ extern const struct bt_conn_auth_cb *bt_auth;
 enum bt_security_err bt_security_err_get(uint8_t hci_err);
 #endif /* CONFIG_BT_SMP || CONFIG_BT_BREDR */
 
+/* Data type to store state related with commnad to be updated
```
command

ppryga replied:
Corrected

joerchan commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -305,6 +313,32 @@ extern const struct bt_conn_auth_cb *bt_auth;
+	state->val = val;
+}
+
+/* Set commnad state for given network buffer */
```
command
```suggestion
/* Set command state related with the command buffer */
```

ppryga replied:
Changed according to suggestion

ppryga commented:
@joerchan thank you for your review. 
Almost all requested comments were implemented. Please check if the code is ok for you now.

ppryga commented:
Just small spelling correction in ull_df.c source file. No other changes added.

ppryga commented:
@carlescufi can you merge the PR?
